apps/server/src/modules/oauth/uc/oauth.uc.ts

```diff
@@ -0,0 +1,137 @@
+import { AuthorizationQuery } from '../controller/dto/authorization.query';
+import { OAuthResponse } from '../controller/dto/oauth.response';
+
+@Injectable()
+export class OauthUc {
+	private logger: ILogger;
+
+	constructor(
+		private readonly systemRepo: SystemRepo,
+		private readonly userRepo: UserRepo,
+		private readonly jwtService: FeathersJwtProvider
+	) {
+		this.logger = new Logger(OauthUc.name);
+	}
+
+	// 0- start Oauth Process
+	async startOauth(query: AuthorizationQuery, systemId: string): Promise<OAuthResponse> {
+		try {
+			// get the authorization code
+			const code: string = this.extractCode(query);
+			// get the Tokens using the authorization token
+			const queryToken: OauthTokenResponse = await this.requestToken(code, systemId);
+			// extract the uuid from the token
+			const uuid = await this.decodeToken(queryToken.id_token);
+			// get the user using the uuid
+			const user: User = await this.findUserById(uuid);
+			// create JWT for the user
+			const jwt: string = await this.getJWTForUser(user);
+			// send response back
+			const HOST = Configuration.get('HOST') as string;
+			const redirectUri = HOST.concat('/dashboard');
+			const response: OAuthResponse = new OAuthResponse({
+				jwt,
+				redirectUri,
+			});
+			return response;
+		} catch (error) {
+			this.logger.log(error);
+		}
+		// send error response back
+		const HOST = Configuration.get('HOST') as string;
+		const redirectUri = HOST.concat('/login');
+		const response: OAuthResponse = new OAuthResponse({
+			redirectUri: `${redirectUri}?error=OauthLoginFailed`,
+		});
+		return response;
+	}
+
+	/**
+	 * @query query input that has either a code or an error
+	 * @return authorization code or throws an error
+	 */
+	extractCode(query: AuthorizationQuery): string {
+		if (query.code) return query.code;
+		if (query.error) throw new Error(query.error);
+		throw new Error('Authorization Query Object has no authorization code or error');
+	}
+
+	mapSystemConfigtoPayload(system: System, code: string): Payload {
+		const payload: Payload = {
+			token_endpoint: system.oauthconfig?.token_endpoint,
+			data: {
+				client_id: system.oauthconfig?.client_id,
+				client_secret: system.oauthconfig?.client_secret,
+				redirect_uri: system.oauthconfig?.token_redirect_uri,
+				grant_type: system.oauthconfig?.grant_type,
+				code,
+			},
+		};
+		if (
+			!(
+				payload.token_endpoint &&
+				payload.data.client_id &&
+				payload.data.client_secret &&
+				payload.data.redirect_uri &&
+				payload.data.grant_type
+			)
+		) {
+			throw new Error('check environment variables');
+		}
+		return payload;
+	}
+
+	// 1- use Authorization Code to get a valid Token
+	async requestToken(code: string, systemId: string) {
+		const system: System = await this.systemRepo.findById(systemId);
+		const payload: Payload = this.mapSystemConfigtoPayload(system, code);
+		const responseToken: AxiosResponse<OauthTokenResponse> = await axios.post(
```

Comment:
axios must be wrapped by an infra service

Reply:
suggestion: create an infra service oauth-client-service which accepts a system or a systems oauthconfig and cares about the required types

Reply:
We used HTTP Service as provided by nest.